tools/e2writeboot.8
.\" e2writeboot - Write secondary SRM bootloader to ext2 filesystem.
.TH e2writeboot 8 "17 January 1995" "David Mosberger" ""
.SH NAME
e2writeboot \- Write secondary SRM bootloader to ext2 filesystem.
.SH SYNOPSIS
\fBe2writeboot\fP <ext2fs> <bootloader>
.SH DESCRIPTION

\fBe2writeboot\fP can be used to write a secondary bootstrap loader to
an ext2 formatted filesystem (normally on a floppy disk).  The loader
is written in a way that allows the SRM firmware to boot from the
filesystem (SRM is the firmware that is also used to boot DEC Unix and
OpenVMS).  Specifically, the loader is written to a physically
contiguous range of sectors and the first sector is updated such that
the SRM firmware can find the boot loader.  The loader is always
stored with filename \fBlinuxboot\fP and if such a file already
exists, it will be overwritten unconditionally.  This command may
fail if the filesystem is highly fragmented.  If so, re-format the
filesystem with \fBmke2fs\fP (this will destroy the existing contents
of the filesystem, so use this with care).

.P
The \fI<ext2fs>\fP argument specifies the filesystem or special device
to which the \fI<bootloader>\fP file should be written to.  The boot
loader file must be an ECOFF object file.  A typical command to make a
floppy disk bootable with the \fIaboot\fP loader is shown below:

.I "e2writeboot /dev/fd0 aboot"

.nf
.SH SEE ALSO
.IR aboot (8) ,
.IR mke2fs (8) ,
.IR swriteboot (8)
.nf
.SH AUTHORS
.nf
Jim Paradis <paradis@amt.tay1.dec.com>
Jay Estabrook <jestabro@amt.tay1.dec.com>
David Mosberger-Tang <davidm@azstarnet.com>
.fi


ChangeLog
2003-03-21  Will Woods <will.woods@hp.com>
	* Added include/string.h, since we have our own string.c, replaced 
	  #includes of <linux/string.h> with "string.h" since linux/string.h
	  no longer has user-includable prototypes.
	  (thanks to Kelledin <kelledin@skarpsey.dyndns.org> for the idea)
	* Changed #includes of our headers to use quotes instead of <>
	* Don't build with -mcpu=ev4 (default code works on ev4 and is smaller)
	* Cleaned up a couple other warnings 

2003-01-14  Will Woods <will.woods@hp.com>
	* Fixed typos in man pages ("True 64")
	* Bumped version number to 0.9b
	* Released 0.9b. Hooray!

2002-09-18  Will Woods <will.woods@hp.com>
	* branched the 0.9 tree to add bugfixes
	* fixed some compilation errors with newer gcc/glibc versions
	* added newer manpages

2001-11-14  Will Woods <will.woods@compaq.com>
        * sdisklabel/sdisklabel.c: fixed my fix to the cylinder counting bug.
          oops. also added (currently undocumented) "force" flag, to skip
          error checking in sdisklabel.

2001-10-26  Will Woods <will.woods@compaq.com>
        * Released aboot 0.9. Yay!

2001-10-25  Will Woods <will.woods@compaq.com>
	* sdisklabel/sdisklabel.c: fixed cylinder-counting bug

2001-10-11  Will Woods <will.woods@compaq.com>
	* applied a patch from the debian version of aboot. Here's the debian
	maintainer's changelog:

  * New upstream (from CVS)
  * I am calling this version of aboot from CVS 0.8.
  * Minor tweeks to make it compile with kernel 2.4.x headers.
  * e2writeboot now returns non-zero on error.  Closes: #107757
  -- Gregory W. Johnson <gjohnson@physics.clarku.edu>  Wed, 15 Aug 2001

	Since he called his version 0.8, I will call this the 0.9 tree, to 
	avoid confusion. Hence this is aboot 0.9pre1. 

2001-10-09  Will Woods <will.woods@compaq.com>
	* aboot.c, disk.c, include/config.c: replaced ABOOT_MINOR and 
	ABOOT_MAJOR with ABOOT_VERSION. Bumped version number to 0.8pre2.
	* aboot.c: fixed ELF header bug that snuck in somewhere.
	* fs/ext2.c, Makefile: fixes for proper #includes to allow compilation
	of userspace testing aboot against the 2.4 kernel.

2001-10-08  Will Woods <will.woods@compaq.com>
	* fs/ext2.c: Fixed elusive overflow bug. If part of the kernel was 
	past the 524,287th block (given 4096 bytes per block) the byte offset
	was too big to fit in an int. 
	* sdisklabel/sdisklabel.c: borrowed code from fdisk for getting
	the correct size of the disk. 
	
2001-07-10  Will Woods <will.woods@compaq.com>
	* aboot.c, tools/objstrip.c: elf_check_arch takes different arguments
	in Linux 2.4 than it did in 2.2; added an aboot_elf_check_arch which
	does the right thing based on version info from linux/version.h. 
	* disk.c: a function called get_options exists in 2.4 kernels, so
	aboot's get_options function was renamed get_aboot_options(). 
	(both changes based on a patch by Christian Groessler <cpg@aladdin.de>)

2001-04-26  Will Woods <will.woods@compaq.com>
	* disk.c: changed /etc/aboot.conf and aboot.conf back to their
	original settings. If it ain't broke, why fix it?
	* added the man pages in doc/man, written by Helge Kreutzmann
	<kreutzm@itp.uni-hannover.de> (Thanks Helge!)

2001-03-20  Will Woods <will.woods@compaq.com>
	* disk.c: swapped position of /etc/aboot.conf and aboot.conf in the 
	configs array, making /aboot.conf the default.
	* isolib.c: added parse_rock_ridge_inode and get_rock_ridge_symlink
	adapted from rock.c in the linux kernel source. 
	* isolib.c: added iso_follow_link and iso_readdir, adapted from
	fs/ext2.c
        * made minor changes here and there to quiet gcc warnings - 
	initializing variables to 0, casting variables properly, etc

2000-06-27  David Huggins-Daines  <dhd@linuxcare.com>

	* sdisklabel/sdisklabel.c (main): Off-by-one error in argument
	checking (not my fault!).
	* disk.c (load_raw): This didn't work at all.  I've changed it to
	use an uncompressed kernel with headers.  Eventually it will be
	fixed so that it can accept compressed kernels as well (by
	checking for gzip magic numbers).
	* lib/vsprintf.c (vsprintf): Patch from Matt Wilson, fix the 'h'
	format in vsprintf().
	* utils.c: Patch from Matt Wilson <msw@redhat.com>, to allow
	compiling against linux 2.3 header files.

2000-05-02  David Huggins-Daines  <dhd@linuxcare.com>

	* Makefile: Added the ability to build a debugging aboot that runs
	as a userspace program.  Great for filesystem code debugging.
	Also, optimize for size instead of speed.
	* aboot.c: Remove a.out support, add multiple ELF pheader
	support.  Also make the parameter offset, and kernel stack offset
	non-hardcoded in order to deal with new, bigger consoles.
	* disk.c: Change "plain" to "uncompressed" and load_aout() to
	load_uncompressed().  Implement multiple pheaders for raw boot
	(untested) and uncompressed boot (tested). Also look for
	alternative config files (needed for CD booting and such)
	* fs/ext2.c: Cache the root inode, and do proper 64-bit arithmetic
	in ext2_iget() so that >2G filesystems work reliably (doh!)
	* head.S: make run_kernel() assembly in order to save a few bytes.
	* include/aboot.h: prototype run_kernel as attribute((noreturn)),
	maybe this will save a few bytes too.
	* fs/ufs.c: Implement dummy fstat().  I don't really know if the
	UFS code still works, and I'm tempted to remove it.
	* include/cons.h: Add a definition of the CTB, though we don't
	really use it yet.
	* include/iso.h: Rock Ridge constants
	* lib/isolib.c: Rock Ridge filenames, code cleanup, and fix the
	bug where 2048-byte blocks don't work.
	* zip/misc.c: necessary stuff for multiple pheaders.

2000-04-10  David Huggins-Daines  <dhuggins@linuxcare.com>

	* lib/isolib.c (iso_iget): Gar!  More stupidity in the ISO code,
	now fixed (see comment about *pnt being overwritten).  With luck
	this should be 0.7 final, so I won't need to do another 0.6

2000-04-02  David Huggins-Daines  <dhuggins@linuxcare.com>

	* lib/isolib.c: Fix some obvious stupidity in the ISO code (could

tools/e2writeboot.c
/* Program to write a primary bootstrap file to a LINUX ext2
 * file system.
 *
 * Usage: e2writeboot fs-image bootfile
 *
 * It is assumed that the "bootfile" is a COFF executable with text,
 * data, and bss contiguous.
 */

#include <grp.h>
#include <pwd.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>
#include <unistd.h>
#include <fcntl.h>
#include <byteswap.h>

#include <sys/stat.h>

#include <e2lib.h>
#include <ext2fs/ext2_fs.h>

#if defined(__linux__)
# include <sys/types.h>
#elif defined(__alpha__) && defined(__osf1__)
  typedef unsigned long		u_int64_t;
#elif defined(__GNUC__)
  typedef unsigned long long	u_int64_t;
#endif

struct  boot_block {
    u_int64_t	vax_code[17];
    u_int64_t	reserved[43];
    u_int64_t	count;
    u_int64_t	lbn;
    u_int64_t	flags;
    u_int64_t	chk_sum;
};

#define CONSOLE_BLOCK_SIZE	512

extern int big_endian;


int
main(int argc, char ** argv)
{
    char		fsname[512];
    char		iobuf[1024];
    char		namebuf[EXT2_NAME_LEN+1];
    struct ext2_inode *	ip;
    int			filesize;
    int			infile;
    int 		blkno;
    int			bootstrap_size;
    int			i;
    int			bs_start;
    char		*bsbuf;
    int			blocksize;
    struct boot_block	*bbp;
    u_int64_t		*lbp, checksum;
    struct stat		st;

    if (argc != 3) {
	printf("Usage: %s ext2-fs input-file\n", argv[0]);
	exit(1);
    }

    strcpy(fsname, argv[1]);
    strcpy(namebuf, "/linuxboot");


    /* "Open" the file system */
    if (ext2_init(fsname, O_RDWR) < 0) {
	exit(1);
    }

    /* Open the input file */
    infile = open(argv[2], 0);
    if (infile < 0) {
	perror(argv[2]);
	ext2_close();
	exit(1);
    }

    /* Figure out just how much data the input file is going
     * to require us to put out. (text+data+bss).
     */
    if (fstat(infile, &st) == -1) {
	perror("fstat");
	ext2_close();
	exit(1);
    }
    blocksize = ext2_blocksize();
    bootstrap_size = st.st_size;
    printf("bootstrap_size: %d -> ", bootstrap_size);
    bootstrap_size += blocksize-1;
    bootstrap_size &= ~(blocksize-1);
    printf("%d\n", bootstrap_size);

    /* Allocate a buffer to hold the entire bootstrap, then read
     * in the text+data segments.
     */
    bsbuf = (char *)malloc(bootstrap_size);
    memset(bsbuf, 0, bootstrap_size);
    read(infile, bsbuf, bootstrap_size);
    close(infile);

    /* Get the inode for the file we want to create */
    ip = ext2_namei(namebuf);
    if(ip) {
	/* This file exists.  Make sure it's a regular file, then
	 * truncate it.
	 */
	if(!S_ISREG(ip->i_mode)) {
	    printf("%s: Not a regular file.  Must remove it first.\n", namebuf);
	    ext2_iput(ip);
	    ext2_close();
	    exit(1);
	}

	printf("Using existing file %s\n", namebuf);
	ext2_truncate(ip);
    }
    else {
	/* Doesn't exist.  Must get the parent directory's inode. */
	char	dirname[EXT2_NAME_LEN+1];
	char	filename[EXT2_NAME_LEN+1];
	struct ext2_inode *dip;
	int	inumber;
	char	*cp;
	int	i;

	strcpy(dirname, namebuf);
	cp = strrchr(dirname, '/');
	if(cp) {
	    *cp = '\0';
	    strcpy(filename, cp+1);
	}
	else {
	    strcpy(filename, dirname);
	    strcpy(dirname, "/");
	}

	dip = ext2_namei(dirname);
	if(!dip) {
	    printf("Directory %s does not exist\n", dirname);
	    ext2_close();
	    exit(1);
	}

	printf("Creating new file %s in directory %s\n", filename, dirname);

	/* Get an inode for the file */
	inumber = ext2_ialloc();
	ip = ext2_iget(inumber);

	if(!ip) {
	    printf("PANIC! ip == NULL\n");
	    exit(1);
	}

	/* Create the directory entry */
	ext2_mknod(dip, filename, inumber);

	/* We're done with the directory for now... */
	ext2_iput(dip);

	/* Set certain fields in the inode (we're not going to get
	 * fancy here... just choose a safe set of defaults...)
	 */
	ip->i_mode = 0550 | S_IFREG;	/* Regular file, r-xr-x--- */
	ip->i_uid = 0;		/* Owned by root */
	ip->i_gid = 0;		/* Group is system */
	ip->i_size = 0;
	ip->i_atime = ip->i_ctime = ip->i_mtime = time(0);
	ip->i_dtime = 0;
	ip->i_links_count = 1;
	ip->i_blocks = 0;
	ip->i_flags = 0;	/* Nothing special */
	ip->osd1.linux1.l_i_version = 0;
	for(i = 0; i < EXT2_N_BLOCKS; i++) {
	    ip->i_block[i] = 0;
	}
	ip->i_generation = 0;
	ip->i_file_acl = 0;
	ip->osd2.linux2.l_i_blocks_hi = 0;
	ip->osd2.linux2.l_i_uid_high = 0;
	ip->osd2.linux2.l_i_gid_high = 0;

    }

    /* At this point we have an inode for an empty regular file.
     * Fill it up!
     */

    bs_start = ext2_fill_contiguous(ip, bootstrap_size/blocksize);
    if(bs_start <= 0) {
	printf("Cannot allocate blocks for %s... goodbye!\n", argv[2]);
	ext2_close();
	exit(1);
    }

    /* Write what we've got out to the file */
    filesize = bootstrap_size;
    blkno = 0;
    while(filesize > 0) {
	ext2_bwrite(ip, blkno, bsbuf+(blkno*blocksize));
	blkno++;
	filesize -= blocksize;
    }

    ip->i_size = bootstrap_size;
    ip->i_mtime = time(0);


    /* Prepare and write out a bootblock */
    memset(iobuf, 0, blocksize);
    bbp = (struct boot_block *)iobuf;

    bbp->count = bootstrap_size / CONSOLE_BLOCK_SIZE;
    bbp->lbn   = (bs_start * blocksize) / CONSOLE_BLOCK_SIZE;
    bbp->flags = 0;

    /* Compute the checksum */
    checksum = 0;
    lbp = (u_int64_t*) bbp;
    for (i = 0; i < CONSOLE_BLOCK_SIZE/8; ++i) {
	checksum += lbp[i];
    }
    bbp->chk_sum = checksum;

    if(big_endian) {
	/* Need to flip the bootblock fields so they come out
	 * right on disk...
	 */
	bbp->count   = (((u_int64_t) bswap_32(bbp->count & 0xffffffff) << 32)
			| bswap_32(bbp->count >> 32));
	bbp->lbn     = (((u_int64_t) bswap_32(bbp->lbn & 0xffffffff) << 32)
			| bswap_32(bbp->lbn >> 32));
	bbp->flags   = (((u_int64_t) bswap_32(bbp->flags & 0xffffffff) << 32)
			| bswap_32(bbp->flags >> 32));
	bbp->chk_sum = (((u_int64_t) bswap_32(bbp->chk_sum & 0xffffffff) << 32)
			| bswap_32(bbp->chk_sum >> 32));
    }

    ext2_write_bootblock((char *) bbp);

    ext2_iput(ip);
    ext2_close();

    printf("%d bytes written to %s\n", bootstrap_size, namebuf);
    return 0;
}
